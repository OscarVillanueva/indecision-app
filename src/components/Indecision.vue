<template>
  <img v-show='img' :src="img" alt="background">
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <input
      type="text"
      name="question"
      id="question"
      placeholder="Realiza una pregunta"
      v-model="question"
    />
    <p>Recuerda terminar con un signo de interrogación (?)</p>

    <div v-show="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>


</template>

<script>
export default {
  name: 'Indecision',
  data() {
    return {
      answer: '',
      img: null,
      isValidQuestion: false,
      question: '',
    }
  },
  methods: {
    async getResponse() {

      try {
        this.answer = 'Pensando . . .'
  
        const { answer, image } = await fetch('https://yesno.wtf/api')
          .then(response => response.json())
  
        switch (answer) {
          case 'yes':
            this.answer = '¡Si!'
            break;
          
          case 'no':
            this.answer = '¡No!'
            break;
        
          default:
            this.answer = 'Tal vez'
            break;
        }
  
        this.img = image
      } catch (error) {
        
        this.img = null
        this.answer = 'No se encontró respuesta'

      }
  
    }
  },
  watch: {
    question(value, oldValue) {
      
      this.isValidQuestion = false

      console.log({ value })

      if(!value.endsWith('?')) return

      this.isValidQuestion = true
      this.getResponse()
      
    } 
  }
}
</script>

<style scoped>
  img, .bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
  }

  .bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
  }

  .indecision-container {
    position: relative;
    z-index: 99;
  }
  
  input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
  }
  input:focus {
    outline: none;
  }

  p {
    color: white;
    font-size: 20px;
    margin-top: 0px;
  }

  h1, h2 {
    color: white;
  }
  
  h2 {
    margin-top: 150px;
  }

</style>