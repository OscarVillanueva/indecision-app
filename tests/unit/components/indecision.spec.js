import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision'

describe('Indecision app', () => {

  let wrapper;
  let consoleSpy;

  // Mock o simulación del fetch y json()
  global.fetch = jest.fn(() => Promise.resolve({
    json: () => Promise.resolve({
      answer: 'yes',
      forced: false,
      image: 'https://yesno.wtf/assets/yes/2.gif'
    })
  }))

  // Antes de cada prueba vuelve a montar el Counter
  beforeEach(() => {
    wrapper = shallowMount(Indecision)
    consoleSpy = jest.spyOn(console, 'log')

    jest.clearAllMocks()
  })

  test('should match with the snapshot', () => expect(wrapper.html()).toMatchSnapshot())

  test('should not occurs anything at write into the input', async () => {

    const getResponseSpy = jest.spyOn(wrapper.vm, 'getResponse')

    const input = wrapper.find('input')
    await input.setValue('Hola mundo')

    expect(consoleSpy).toHaveBeenCalled()
    expect(getResponseSpy).not.toHaveBeenCalled()

  })

  test('should do a fetch at write an ?', async () => {

    const getResponseSpy = jest.spyOn(wrapper.vm, 'getResponse')

    const input = wrapper.find('input')
    await input.setValue('Hola mundo?')

    expect(getResponseSpy).toHaveBeenCalled()

  })

  test('tests in getResponse', async () => {

    await wrapper.vm.getResponse()

    expect(wrapper.find('img').isVisible()).toBeTruthy()
    expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
    expect(wrapper.vm.answer).toBe('¡Si!')

  })

  test('should catch the error in API', async () => {

    global.fetch.mockImplementationOnce(() => Promise.reject('API is DOWN'))

    await wrapper.vm.getResponse()

    expect(wrapper.find('img').isVisible()).toBeFalsy()
    expect(wrapper.vm.answer).toBe('No se encontró respuesta')

  })

})