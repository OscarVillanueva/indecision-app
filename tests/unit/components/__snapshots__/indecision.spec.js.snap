// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision app should match with the snapshot 1`] = `
<img alt="background" style="display: none;">
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" name="question" id="question" placeholder="Realiza una pregunta">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <div style="display: none;">
    <h2></h2>
    <h1></h1>
  </div>
</div>
`;
